Library/OcCpuLib/OcCpuLib.c

```diff
@@ -1003,110 +1003,113 @@ ScanIntelProcessor (
-    Cpu->CurBusRatio = (UINT8) (PerfStatus.Bits.State >> 8U);
-    PlatformInfo.Uint64 = AsmReadMsr64 (MSR_NEHALEM_PLATFORM_INFO);
-    Cpu->MinBusRatio = (UINT8) PlatformInfo.Bits.MaximumEfficiencyRatio;
-    Cpu->MaxBusRatio = (UINT8) PlatformInfo.Bits.MaximumNonTurboRatio;
-  } else if (Cpu->Model >= CPU_MODEL_PENRYN) {
-    PerfStatus.Uint64 = AsmReadMsr64 (MSR_IA32_PERF_STATUS);
-    Cpu->MaxBusRatio = (UINT8) (PerfStatus.Uint64 >> 8U) & 0x1FU;
-    //
-    // Undocumented values:
-    // Non-integer bus ratio for the max-multi.
-    // Non-integer bus ratio for the current-multi.
-    //
-    // MaxBusRatioDiv = (UINT8)(PerfStatus.Uint64 >> 46U) & 0x01U;
-    // CurrDiv = (UINT8)(PerfStatus.Uint64 >> 14U) & 0x01U;
-    //
-  }
+  // When the CPU is visualized and cpuid invtsc is enabled, then we already get the information
+  // we want outside the function, skip anyway.
+  // Things maybe difference in other hypervisors, but should work with QEMU/VMWare now
+  if (!(Cpu->Hypervisor && (Cpu->Features & CPUID_EXTFEATURE_TSCI))){
+      // TODO: this may not be accurate on some older processors.
+      //
+      if (Cpu->Model >= CPU_MODEL_NEHALEM) {
+        PerfStatus.Uint64 = AsmReadMsr64 (MSR_IA32_PERF_STATUS);
+        Cpu->CurBusRatio = (UINT8) (PerfStatus.Bits.State >> 8U);
+        PlatformInfo.Uint64 = AsmReadMsr64 (MSR_NEHALEM_PLATFORM_INFO);
+        Cpu->MinBusRatio = (UINT8) PlatformInfo.Bits.MaximumEfficiencyRatio;
+        Cpu->MaxBusRatio = (UINT8) PlatformInfo.Bits.MaximumNonTurboRatio;
+      } else if (Cpu->Model >= CPU_MODEL_PENRYN) {
+        PerfStatus.Uint64 = AsmReadMsr64 (MSR_IA32_PERF_STATUS);
+        Cpu->MaxBusRatio = (UINT8) (PerfStatus.Uint64 >> 8U) & 0x1FU;
+        //
+        // Undocumented values:
+        // Non-integer bus ratio for the max-multi.
+        // Non-integer bus ratio for the current-multi.
+        //
+        // MaxBusRatioDiv = (UINT8)(PerfStatus.Uint64 >> 46U) & 0x01U;
+        // CurrDiv = (UINT8)(PerfStatus.Uint64 >> 14U) & 0x01U;
+        //
+      }
 
-  if (Cpu->Model >= CPU_MODEL_NEHALEM
-    && Cpu->Model != CPU_MODEL_NEHALEM_EX
-    && Cpu->Model != CPU_MODEL_WESTMERE_EX) {
-    TurboLimit.Uint64 = AsmReadMsr64 (MSR_NEHALEM_TURBO_RATIO_LIMIT);
-    Cpu->TurboBusRatio1 = (UINT8) TurboLimit.Bits.Maximum1C;
-    Cpu->TurboBusRatio2 = (UINT8) TurboLimit.Bits.Maximum2C;
-    Cpu->TurboBusRatio3 = (UINT8) TurboLimit.Bits.Maximum3C;
-    Cpu->TurboBusRatio4 = (UINT8) TurboLimit.Bits.Maximum4C;
-  }
+      if (Cpu->Model >= CPU_MODEL_NEHALEM
+        && Cpu->Model != CPU_MODEL_NEHALEM_EX
+        && Cpu->Model != CPU_MODEL_WESTMERE_EX) {
+        TurboLimit.Uint64 = AsmReadMsr64 (MSR_NEHALEM_TURBO_RATIO_LIMIT);
+        Cpu->TurboBusRatio1 = (UINT8) TurboLimit.Bits.Maximum1C;
+        Cpu->TurboBusRatio2 = (UINT8) TurboLimit.Bits.Maximum2C;
+        Cpu->TurboBusRatio3 = (UINT8) TurboLimit.Bits.Maximum3C;
+        Cpu->TurboBusRatio4 = (UINT8) TurboLimit.Bits.Maximum4C;
```

Comment:
I personally believe it can be left set to 0 (we have ZeroMem, so we do not need to explicitly set the fields). These values do not seem to make sense on hyoervisors, so just nor setting them is fine to me.